// distube/dist/core/DisTubeBase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DisTubeBase = void 0;
/** @private */
class DisTubeBase {
    constructor(distube) {
        /**
         * DisTube
         * @type {DisTube}
         * @private
         */
        this.distube = distube;
    }
    /**
     * Emit the {@link DisTube} of this base
     * @private
     * @param {string} eventName Event name
     * @param {...any} args arguments
     * @returns {boolean}
     */
    emit(eventName, ...args) {
        return this.distube.emit(eventName, ...args);
    }
    /**
     * Emit error event
     * @param {Error} error error
     * @param {Discord.TextChannel?} channel Text channel where the error is encountered.
     * @private
     */
    emitError(error, channel) {
        this.distube.emitError(error, channel);
    }
    /**
     * The queue manager
     * @type {QueueManager}
     * @readonly
     * @private
     */
    get queues() {
        return this.distube.queues;
    }
    /**
     * The voice manager
     * @type {DisTubeVoiceManager}
     * @readonly
     * @private
     */
    get voices() {
        return this.distube.voices;
    }
    /**
     * Discord.js client
     * @type {Discord.Client}
     * @readonly
     * @private
     */
    get client() {
        return this.distube.client;
    }
    /**
     * DisTube options
     * @type {DisTubeOptions}
     * @readonly
     * @private
     */
    get options() {
        return this.distube.options;
    }
    /**
     * DisTube handler
     * @type {DisTubeHandler}
     * @readonly
     * @private
     */
    get handler() {
        return this.distube.handler;
    }
}
exports.DisTubeBase = DisTubeBase;
exports.default = DisTubeBase;
