// ards-client/tags/hentai.js
const requestHandler = require("ards-client/tags/handler");
class Hentai {
    neko() {
        let konaOrder = ['id', 'id_desc', 'score', 'score_asc', 'mpixels', 'mpixels_asc', 'landscape', 'portrait']
        let urls = [
            {url: `https://danbooru.donmai.us/posts.json?limit=120&order=none&tags=rating:e cat_girl`, source: "danbooru"}, 
            {url: `https://konachan.com/post.json?limit=120&tags=rating:e catgirl&order=${konaOrder[Math.floor(Math.random() * konaOrder.length)]}`, source: "konachan"},
            {url: `https://yande.re/post.json?limit=120&tags=rating:e nekomimi`, source: "yandere"}
        ];
        let data = urls[Math.floor(Math.random() *urls.length)];
        return requestHandler.makeRequest(data.source, data.url);
    }
    danbooru(tags) {
        let url = `https://danbooru.donmai.us/posts.json?limit=120&order=none&tags=${tags}`;
        return requestHandler.makeRequest("danbooru", url);
    }
    e621(tags, useragent) {
        let url = `https://e621.net/posts.json?limit=240&order=none&tags=${tags.join(' ')}`;
        return requestHandler.makeRequest("e621", url, useragent);
    }
    konachan(tags) {
        let order = ['id', 'id_desc', 'score', 'score_asc', 'mpixels', 'mpixels_asc', 'landscape', 'portrait']
        let url = `https://konachan.com/post.json?limit=120&tags=${tags.join(' ')}&order=${order[Math.floor(Math.random() * order.length)]}`
        return requestHandler.makeRequest("konachan", url);
    }
    rule34(tags) {
        let sort = ['id', 'score', 'rating', 'user', 'height', 'parent', 'source', 'updated']
        let url = `https://rule34.xxx/index.php?page=dapi&s=post&q=index&limit=120&${sort}&tags=${tags.join(' ')}`
        return requestHandler.makeRequest("rule34", url);
    }
    yandere(tags) {
        let url = `https://yande.re/post.json?limit=120&tags=${tags.join(' ')}`;
        return requestHandler.makeRequest("yandere", url);
    }
}
module.exports = new Hentai();
